transfer/custom.go

```diff
@@ -0,0 +1,382 @@
+	}
+
+	customCtx, ok := ctx.(*customAdapterWorkerContext)
+	if !ok {
+		return fmt.Errorf("Context object for custom transfer %q was of the wrong type", a.name)
+	}
+	var authCalled bool
+
+	rel, ok := t.Object.Rel(a.getOperationName())
+	if !ok {
+		return errors.New("Object not found on the server.")
+	}
+	var req *customAdapterTransferRequest
+	if a.direction == Upload {
+		req = NewCustomAdapterUploadRequest(t.Object.Oid, t.Object.Size, t.Path, rel)
+	} else {
+		req = NewCustomAdapterDownloadRequest(t.Object.Oid, t.Object.Size, rel)
+	}
+	err := a.sendMessage(customCtx, req)
+	if err != nil {
+		return err
+	}
+
+	// 1..N replies (including progress & one of download / upload)
+	var complete bool
+	for !complete {
+		resp, err := a.readResponse(customCtx)
+		if err != nil {
+			return err
+		}
+		var wasAuthOk bool
+		switch resp.Id {
+		case "progress":
+			// Progress
+			if resp.Oid != t.Object.Oid {
+				return fmt.Errorf("Unexpected oid %q in response, expecting %q", resp.Oid, t.Object.Oid)
+			}
+			if cb != nil {
+				cb(t.Name, t.Object.Size, resp.BytesSoFar, resp.BytesSinceLast)
+			}
+			wasAuthOk = resp.BytesSoFar > 0
+		case "complete":
+			// Download/Upload complete
+			if resp.Oid != t.Object.Oid {
+				return fmt.Errorf("Unexpected oid %q in response, expecting %q", resp.Oid, t.Object.Oid)
+			}
+			if resp.Error != nil {
+				return fmt.Errorf("Error transferring %q: %v", t.Object.Oid, resp.Error)
+			}
+			if a.direction == Download {
+				// So we don't have to blindly trust external providers, check SHA
+				if err = tools.VerifyFileHash(t.Object.Oid, resp.Path); err != nil {
+					return fmt.Errorf("Downloaded file failed checks: %v", err)
+				}
+				// Move file to final location
+				if err = tools.RenameFileCopyPermissions(resp.Path, t.Path); err != nil {
+					return fmt.Errorf("Failed to copy downloaded file: %v", err)
+				}
+			} else if a.direction == Upload {
+				if err = api.VerifyUpload(t.Object); err != nil {
+					return err
+				}
+			}
+			wasAuthOk = true
+			complete = true
+		}
+		// Fall through from both progress and completion messages
+		// Call auth on first progress or success to free up other workers
+		if wasAuthOk && authOkFunc != nil && !authCalled {
+			authOkFunc()
+			authCalled = true
+		}
+	}
+
+	return nil
+}
+
+func newCustomAdapter(name string, dir Direction, path, args string, concurrent bool) *customAdapter {
+	c := &customAdapter{newAdapterBase(name, dir, nil), path, args, concurrent, 3}
+	// self implements impl
+	c.transferImpl = c
+	return c
+}
+
+// Initialise custom adapters based on current config
+func ConfigureCustomAdapters() {
+	pathRegex := regexp.MustCompile(`lfs.customtransfer.([^.]+).path`)
+	for k, v := range config.Config.AllGitConfig() {
+		if match := pathRegex.FindStringSubmatch(k); match != nil {
+			name := match[1]
+			path := v
+			var args string
+			var concurrent bool
+			var direction string
+			// retrieve other values
```

Comment:
I would change the operation on the substring match, so it continues if there's no match. This reduces the level of indentation. 

Also, it wasn't immediately clear what the 3 vars are for, until we got to the closure below. I think a quick comment would be helpful, since vars are unusual.